activesupport/lib/active_support/message_encryptor.rb

```diff
@@ -90,8 +93,12 @@ def encrypt_and_sign(value)
 
     # Decrypt and verify a message. We need to verify the message in order to
     # avoid padding attacks. Reference: http://www.limited-entropy.com/padding-oracle-attacks.
-    def decrypt_and_verify(value)
-      _decrypt(verifier.verify(value))
+    def decrypt_and_verify(value, purpose = nil)
```

Comment:
In fact let's wrap the payload metadata handling in an `ActiveSupport::Messages::Metadata` class. Here's the rough structure for it (as I see it):

```ruby
# active_support/messages/metadata.rb
module ActiveSupport
  module Messages
    class Metadata # :nodoc:
      def initialize(expires_at, purpose)
        @expires_at, @purpose = expires_at, purpose
      end

      class << self
        def wrap(message, expires_at: nil, expires_in: nil, purpose: nil)
          { "value" => message, "_rails" => { "exp" => pick_expiry(expires_at, expires_in), "pur" => purpose } }
        end

        def verify(message, purpose)
          metadata = extract_metadata(message)
          message["value"] if metadata.nil? || (metadata.match?(purpose) && metadata.fresh?)
        end

        private
          def pick_expiry(expires_at, expires_in)
            # Needs to use iso8601(3) and have expires_at take precedence. See SignedGlobalID and Active Storage for inspiration.
          end

          def extract_metadata(message)
            # Check if the metadata came from us and create a new instance.
          end
      end

      def match?(purpose)
      end

      def fresh?
      end
    end
  end
end
```

We'll need to use string keys throughout such that messages encoded with JSON will work when decoded. We can then use this class in both the verifier and encryptor.

The `_rails` key stores the metadata and lets us handle backwardscompatibility. E.g. if that key is missing we can't provide the extra metadata verification, so we should just skip it.